include/ipr/interface

```diff
@@ -305,6 +305,7 @@ namespace ipr {
       using Location_span = std::pair<Unit_location, Unit_location>;
       virtual const Location_span& span() const = 0;
       virtual const Region& enclosing() const = 0;
+      virtual const Expr& owner() const = 0;
       virtual const Sequence<Expr>& body() const = 0;
```

Comment:
I am wondering if Region can be ownerless. What if Region always have to have an owner, maybe the issue is in the Callable_species. Maybe Callable_species should not host a Parameter_list at all, instead, it would refer to a Parameter_list of the associated Fundecl?

Reply:
> Maybe Callable_species should not host a Parameter_list at all, instead, it would refer to a Parameter_list of the associated Fundecl?

Wouldn't that assume that there is a `Fundecl` that would own the `Parameter_list`?  A species hasn't been elaborated yet, so there is no `Decl` for it.

Reply:
Let me push back against Optional Owner() a bit more. I think it can be well defined for cxx::form as well. It will be pointing at the Specifiers_spread directive. 

Note that Callable_species hosts Parameter_list, even if the target (Fundecl) is assigned to Callable_species, Fundecl will have fully independent Parameter_list.

Thus we don't have a case when we won't have an owner.

Reply:
See the updated region-owner.cxx which now correctly sets up callable_species region owner

Reply:
> Let me push back against Optional Owner() a bit more. I think it can be well defined for cxx::form as well. It will be pointing at the Specifiers_spread directive.

No, the goal isn't just to look around and pick anything that looks like an `Expr` and use that as `owner()`.

Furthermore, there seems to be an assumption that a `Callable_species` will always lead to a `Fundecl`, when elaborated.  That is not the case.
 Consider

```
int i, (*f)(int (*pf)(double x));
```

The `Callable_species` `(*f)(int (*pf)(double x))` does not lead to a `Fundecl`.

> 
> Note that `Callable_species` hosts Parameter_list, even if the target (`Fundecl`) is assigned to Callable_species, `Fundecl` will have fully independent `Parameter_list`.

I don't follow.  Could you expand on the steps based on the above example?  Why would there be a fully independent `Parameter_list`?

> 
> Thus we don't have a case when we won't have an owner.

You're going too fast for me :-)

Reply:
I agree that the target for callable_species is not always a fundecl. Let me withdraw that line of argument it was not fundamental assumption and was a distraction.

I want to recover the property we used to have that the Region always have an owner. Callable_species for a second made me doubt this property, but, after further reflection, I think it is possible to recover that property.

I would like us to explore if we can get there.

Reply:
> I want to recover the property we used to have that the Region always have an owner. 

As I explained in our in-person conversation, that proved out to be _one_ of the flaws that led me to remove `owner()` in the first place.  I believe we need to accept that it is optional, especially when non-generative entities are involved and in some cases of homogeneous scopes.  Here is a comment I wrote internally as I was exploring a different route:

```
  // of expressions or directives.  It is mostly useful for capturing the notion of
   // declarative region, and of scope (in Standard C++ sense).  As such, a generative
   // entity that can contain declarations has a region: that region is said to be 
   // owned by that entity.  On the hand, not all regions are associcated with 
   // generative entities. For example, in
   //         int i, (*pf)(int (*phi)(double x));
   // the entity (type) `int (*)(double)` is not generative (types are unified!).  Similarly,
   // the entity (type) `int (*)(int (*)(double))` is not generative.  The region associated
   // with the
   //
  ```

> Callable_species for a second made me doubt this property, but, after further reflection, I think it is possible to recover that property.

Even without `Callable_species`, the property as state does not hold.

Reply:
Okay. Owner() is now Optional<Expr>